src/Lecture2.hs

```diff
@@ -243,53 +348,97 @@ Add (Var "x") (Lit 1)
 Add (Var "y") (Add (Var "x") (Lit 10))
 -}
 
-{- | We want to evaluate such expressions. We can associate a value
-with a variable using a list of pairs.
-
-You can use the @lookup@ function to search in this list by a variable name:
-
- * https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:lookup
--}
+-- | We want to evaluate such expressions. We can associate a value
+-- with a variable using a list of pairs.
+--
+-- You can use the @lookup@ function to search in this list by a variable name:
+--
+-- * https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:lookup
 type Variables = [(String, Int)]
 
-{- | Unfortunately, it's not guaranteed that variables in our @Expr@
-data type are present in the given list. So we're going to introduce a
-separate data for possible evaluation errors.
-
-Normally, this would be a sum type with several constructors
-describing all possible errors. But we have only one error in our
-evaluation process.
--}
+-- | Unfortunately, it's not guaranteed that variables in our @Expr@
+-- data type are present in the given list. So we're going to introduce a
+-- separate data for possible evaluation errors.
+--
+-- Normally, this would be a sum type with several constructors
+-- describing all possible errors. But we have only one error in our
+-- evaluation process.
 data EvalError
-    = VariableNotFound String
-    deriving (Show, Eq)
-
-{- | Having all this set up, we can finally implement an evaluation function.
-It returns either a successful evaluation result or an error.
--}
-eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
-{- | Compilers also perform optimizations! One of the most common
-optimizations is "Constant Folding". It performs arithmetic operations
-on all constants known during compile time. This way you can write
-more verbose and clear code that works as efficient as its shorter
-version.
+  = VariableNotFound String
+  deriving (Show, Eq)
 
-For example, if you have an expression:
+type EvalResult = Either EvalError Int
 
-x + 10 + y + 15 + 20
-
-The result of constant folding can be:
-
-x + y + 45
+-- | Having all this set up, we can finally implement an evaluation function.
+-- It returns either a successful evaluation result or an error.
+eval :: Variables -> Expr -> Either EvalError Int
+eval vars expr = case expr of
+  Lit cons -> Right cons
+  Var var -> case lookup var vars of
+    Nothing -> Left (VariableNotFound var)
+    Just val -> Right val
+  Add ea eb -> evalAdd (eval vars ea) (eval vars eb)
+    where
+      evalAdd :: EvalResult -> EvalResult -> EvalResult
+      evalAdd (Left err) _ = Left err
+      evalAdd _ (Left err) = Left err
+      evalAdd (Right ra) (Right rb) = Right (ra + rb)
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️